File: streams/library/cluster/profiler.cpp
/*
 * Copyright 2006 by the Massachusetts Institute of Technology.
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

#include "profiler.h"

#include <stdlib.h>
#include <stdio.h>
#include <assert.h>

// counts of all operations
int profiler::float_total = 0;
int profiler::int_total = 0;

// counts of specific operations
int* profiler::float_ops = (int*)calloc(NUM_OPCODES(), sizeof(int));
int* profiler::int_ops = (int*)calloc(NUM_OPCODES(), sizeof(int));

// names of operations
char** profiler::OP_TO_NAME = (char**)calloc(NUM_OPCODES(), sizeof(char*));

// for constructing operations
int profiler::curr_opcode = 0;
int profiler::register_op(char* name) {
  assert(curr_opcode < NUM_OPCODES()); // increase NUM_OPCODES if fails
  int result = curr_opcode++;
  OP_TO_NAME[result] = name;
  return result;
}

// initialized by set_num_ids
int* profiler::id_counts;
int profiler::num_ids;

// the names of these fields should match those targetted from the
// compiler
int profiler::BINOP_ADD =       register_op("add");
int profiler::BINOP_SUB =       register_op("sub");
int profiler::BINOP_MUL =       register_op("mul");
int profiler::BINOP_DIV =       register_op("div");
int profiler::BINOP_MOD =       register_op("mod");
int profiler::BINOP_AND =       register_op("and");
int profiler::BINOP_OR =        register_op("or");
int profiler::BINOP_EQ =        register_op("eq");
int profiler::BINOP_NEQ =       register_op("neq");
int profiler::BINOP_LT =        register_op("lt");
int profiler::BINOP_LE =        register_op("le");
int profiler::BINOP_GT =        register_op("gt");
int profiler::BINOP_GE =        register_op("ge");
// These are bitwise AND/OR/XOR:       
int profiler::BINOP_BAND =      register_op("band");
int profiler::BINOP_BOR =       register_op("bor");
int profiler::BINOP_BXOR =      register_op("bxor");
int profiler::BINOP_LSHIFT =    register_op("lshift");
int profiler::BINOP_RSHIFT =    register_op("rshift");
// unary ops
int profiler::UNOP_NOT =        register_op("unary_not");
// not sure what a unary positive really is, but kopi defines it, so
// put it here just in case
int profiler::UNOP_POS =        register_op("unary_pos");
int profiler::UNOP_NEG =        register_op("unary_neg");
int profiler::UNOP_PREINC =     register_op("preinc");
int profiler::UNOP_POSTINC =    register_op("postinc");
int profiler::UNOP_PREDEC =     register_op("predec");
int profiler::UNOP_POSTDEC =    register_op("postdec");
int profiler::UNOP_COMPLEMENT = register_op("complement");
// these are math functions:           
int profiler::FUNC_ABS =        register_op("abs");
int profiler::FUNC_ACOS =       register_op("cos");
int profiler::FUNC_ASIN =       register_op("sin");
int profiler::FUNC_ATAN =       register_op("tan");
int profiler::FUNC_ATAN2 =      register_op("tan2");
int profiler::FUNC_CEIL =       register_op("ceil");
int profiler::FUNC_COS =        register_op("cos");
int profiler::FUNC_SIN =        register_op("sin");
int profiler::FUNC_COSH =       register_op("cosh");
int profiler::FUNC_SINH =       register_op("sinh");
int profiler::FUNC_EXP =        register_op("exp");
int profiler::FUNC_FABS =       register_op("abs");
int profiler::FUNC_MODF =       register_op("modf");
int profiler::FUNC_FMOD =       register_op("mod");
int profiler::FUNC_FREXP =      register_op("frexp");
int profiler::FUNC_FLOOR =      register_op("floor");
int profiler::FUNC_LOG =        register_op("log");
int profiler::FUNC_LOG10 =      register_op("log10");
int profiler::FUNC_POW =        register_op("pow");
int profiler::FUNC_ROUND =      register_op("round");
int profiler::FUNC_RINT =       register_op("int");
int profiler::FUNC_SQRT =       register_op("sqrt");
int profiler::FUNC_TANH =       register_op("tanh");
int profiler::FUNC_TAN =        register_op("tan");
// make sure you set this to be the number of operations
int profiler::NUM_OPCODES() { return 50; }

/**
 * One of the main callbacks from the instrumented StreamIt code.
 * This version is for int values.
 *
 * @param op      The operation being performed.
 * @param id      A static identifier for the line of code calling this.
 * @param val     One of the values taking part in the operation.
 *
 * @return The value of the parameter <val>.
 */
int profiler::register_op(int op, int id, int val) {
  id_counts[id]++;
  int_ops[op]++;
  int_total++;
  return val;
}

/**
 * One of the main callbacks from the instrumented StreamIt code.
 * This version is for float values.
 *
 * @param op      The operation being performed.
 * @param id      A static identifier for the line of code calling this.
 * @param val     One of the values taking part in the operation.
 *
 * @return The value of the parameter <val>.
 */
float profiler::register_op(int op, int id, float val) {
  id_counts[id]++;
  float_ops[op]++;
  float_total++;
  return val;
}

/**
 * Called before execution to tell the profiler how many operation
 * ID's there will be.
 */
void profiler::set_num_ids(int _num_ids) {
  id_counts = (int*)calloc(_num_ids, sizeof(int));
  num_ids = _num_ids;
}

/**
 * Called when the program has finished, to print results, etc.
 */
void profiler::summarize() {
  FILE *f = fopen("./countops.c.log", "w");

  // total ops 
  int ops_total = float_total + int_total;
  fprintf(f, "Total ops:   %d\n", ops_total);
  fprintf(f, "  float ops: %d\n", float_total);
  fprintf(f, "  int ops:   %d\n", int_total);

  /* NOT IMPLEMENTED YET

  // communication
  fprintf(f, "\n");
  fprintf(f, "Items pushed: %d\n", push_total);

  // communication-to-computation ratio
  fprintf(f, "\n");
  fprintf(f, "Ops / push:   %f\n", (ops_total / (float)push_total));

  */

  // float ops summary
  fprintf(f, "\nFloat ops breakdown:\n");
  for (int i=0; i<NUM_OPCODES(); i++) {
    fprintf(f, "  %s: %d\n", OP_TO_NAME[i], float_ops[i]);
  }
  // int ops summary
  fprintf(f, "\nInt ops breakdown:\n");
  for (int i=0; i<NUM_OPCODES(); i++) {
    fprintf(f, "  %s: %d\n", OP_TO_NAME[i], int_ops[i]);
  }

  // the count by actual statement in the code
  fprintf(f, "\nCount for each static operation ID (see .cpp files for ID's):\n");
  for (int i=0; i<num_ids; i++) {
    float percent = 100*id_counts[i]/(float)ops_total;
    fprintf(f, "  %d: %d (%2.2f\%)\n", i, id_counts[i], percent);
  }

  // the sorted count by actual statement in the code
  fprintf(f, "\nSorted count for each static operation ID (see .cpp files for ID's):\n");
  // just do selection sort because I don't know C++ sort libraries
  for (int i=0; i<num_ids; i++) {
    int max = 0;
    // find greatest
    for (int j=0; j<num_ids; j++) {
      if (id_counts[j] > id_counts[max]) {
	max = j;
      }
    }
    // print greatest
    float percent = 100*id_counts[max]/(float)ops_total;
    fprintf(f, "  %d: %d (%2.2f\%)\n", max, id_counts[max], percent);
    // zero-out greatest
    id_counts[max] = -1;
  }

  fclose(f);

  printf("Profile information written to countops.c.log.\n");
}

File: streams/src/streamit/library/Profiler.java
/*
 * Copyright 2003 by the Massachusetts Institute of Technology.
 *
 * Permission to use, copy, modify, and distribute this
 * software and its documentation for any purpose and without
 * fee is hereby granted, provided that the above copyright
 * notice appear in all copies and that both that copyright
 * notice and this permission notice appear in supporting
 * documentation, and that the name of M.I.T. not be used in
 * advertising or publicity pertaining to distribution of the
 * software without specific, written prior permission.
 * M.I.T. makes no representations about the suitability of
 * this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 */

package streamit.library;

import java.io.FileOutputStream;
import java.io.PrintStream;
import java.io.IOException;
import java.util.List;
import java.util.Set;
import java.util.HashSet;
import java.util.LinkedList;

/**
 * This class can tabulate profiling statistics for execution of
 * StreamIt programs in the Java library.  It is notified of
 * arithmetic operations by instrumentation code that is inserted by
 * the frontend (when run with strc -countops -library).
 */
public class Profiler {
    // Temporary list for constructing names of operations.
    private static final List<String> tempOpToName = new LinkedList<String>();
    // for constructing operations
    private static final int registerOp(String name) {
        tempOpToName.add(name);
        return tempOpToName.size()-1;
    }

    // the names of these fields should match those targetted from
    // streamit.frontend.tojava.NodesToJava.  
    public static final int BINOP_ADD =       registerOp("add");
    public static final int BINOP_SUB =       registerOp("sub");
    public static final int BINOP_MUL =       registerOp("mul");
    public static final int BINOP_DIV =       registerOp("div");
    public static final int BINOP_MOD =       registerOp("mod");
    public static final int BINOP_AND =       registerOp("and");
    public static final int BINOP_OR =        registerOp("or");
    public static final int BINOP_EQ =        registerOp("eq");
    public static final int BINOP_NEQ =       registerOp("neq");
    public static final int BINOP_LT =        registerOp("lt");
    public static final int BINOP_LE =        registerOp("le");
    public static final int BINOP_GT =        registerOp("gt");
    public static final int BINOP_GE =        registerOp("ge");
    // These are bitwise AND/OR/XOR:       
    public static final int BINOP_BAND =      registerOp("band");
    public static final int BINOP_BOR =       registerOp("bor");
    public static final int BINOP_BXOR =      registerOp("bxor");
    public static final int BINOP_LSHIFT =    registerOp("lshift");
    public static final int BINOP_RSHIFT =    registerOp("rshift");
    // unary ops
    public static final int UNOP_NOT =        registerOp("unary_not");
    // not sure what a unary positive really is, but kopi defines it,
    // so put it here for consistency
    public static final int UNOP_POS =        registerOp("unary_pos");
    public static final int UNOP_NEG =        registerOp("unary_neg");
    public static final int UNOP_PREINC =     registerOp("preinc");
    public static final int UNOP_POSTINC =    registerOp("postinc");
    public static final int UNOP_PREDEC =     registerOp("predec");
    public static final int UNOP_POSTDEC =    registerOp("postdec");
    public static final int UNOP_COMPLEMENT = registerOp("complement");
    // these are math functions:           
    public static final int FUNC_ABS =        registerOp("abs");
    public static final int FUNC_ACOS =       registerOp("cos");
    public static final int FUNC_ASIN =       registerOp("sin");
    public static final int FUNC_ATAN =       registerOp("tan");
    public static final int FUNC_ATAN2 =      registerOp("tan2");
    public static final int FUNC_CEIL =       registerOp("ceil");
    public static final int FUNC_COS =        registerOp("cos");
    public static final int FUNC_SIN =        registerOp("sin");
    public static final int FUNC_COSH =       registerOp("cosh");
    public static final int FUNC_SINH =       registerOp("sinh");
    public static final int FUNC_EXP =        registerOp("exp");
    public static final int FUNC_FABS =       registerOp("abs");
    public static final int FUNC_MODF =       registerOp("modf");
    public static final int FUNC_FMOD =       registerOp("mod");
    public static final int FUNC_FREXP =      registerOp("frexp");
    public static final int FUNC_FLOOR =      registerOp("floor");
    public static final int FUNC_LOG =        registerOp("log");
    public static final int FUNC_LOG10 =      registerOp("log10");
    public static final int FUNC_POW =        registerOp("pow");
    public static final int FUNC_RANDOM =     registerOp("random");
    public static final int FUNC_ROUND =      registerOp("round");
    public static final int FUNC_RINT =       registerOp("int");
    public static final int FUNC_SQRT =       registerOp("sqrt");
    public static final int FUNC_TANH =       registerOp("tanh");
    public static final int FUNC_TAN =        registerOp("tan");
    public static final int FUNC_MIN =        registerOp("minf");
    public static final int FUNC_MAX =        registerOp("maxf");

    // construct array of names
    public static final String[] OP_TO_NAME = tempOpToName.toArray(new String[0]);
    // number of operations
    public static final int NUM_OPS = OP_TO_NAME.length;

    // counts of all operations
    private static long floatTotal = 0;
    private static long intTotal = 0;
    private static long boolTotal = 0;

    // counts of all communication
    private static long pushTotal = 0;
    private static long popTotal = 0;

    // counts of specific operations
    private static long[] floatOps = new long[NUM_OPS];
    private static long[] intOps = new long[NUM_OPS];
    private static long[] boolOps = new long[NUM_OPS];

    // counts of how many times each operation ID executed.  There is
    // an ID assigned to each static arithmetic operation by the
    // frontend.  This is similar to a line number, except that there
    // could be multiple operation ID's per line.
    static long[] idCounts;
    // the actual line of code that is being computed at a given id
    static String[] idCode;

    /**
     * One of the main callbacks from the instrumented StreamIt code.
     * This version is for int values.
     *
     * @param op      The operation being performed.
     * @param id      A static identifier for the line of code calling this.
     * @param val     One of the values taking part in the operation.
     * @param code    The code that computes this op.
     *
     * @return The value of the parameter <pre>val</pre>.
     */
    public static int registerOp(int op, int id, String code, int val) {
        idCode[id] = code;
        idCounts[id]++;
        intOps[op]++;
        intTotal++;
        return val;
    }

    /**
     * One of the main callbacks from the instrumented StreamIt code.
     * This version is for float values.
     *
     * @param op      The operation being performed.
     * @param id      A static identifier for the line of code calling this.
     * @param val     One of the values taking part in the operation.
     * @param code    The code that computes this op.
     *
     * @return The value of the parameter <pre>val</pre>.
     */
    public static float registerOp(int op, int id, String code, float val) {
        idCode[id] = code;
        idCounts[id]++;
        floatOps[op]++;
        floatTotal++;
        return val;
    }

    /**
     * One of the main callbacks from the instrumented StreamIt code.
     * This version is for boolean values.
     *
     * @param op      The operation being performed.
     * @param id      A static identifier for the line of code calling this.
     * @param val     One of the values taking part in the operation.
     * @param code    The code that computes this op.
     *
     * @return The value of the parameter <pre>val</pre>.
     */
    public static boolean registerOp(int op, int id, String code, boolean val) {
        idCode[id] = code;
        idCounts[id]++;
        boolOps[op]++;
        boolTotal++;
        return val;
    }

    /**
     * Called before execution to tell the profiler how many operation
     * ID's there will be.
     */
    public static void setNumIds(int numIds) {
        // initialize arrays
        idCounts = new long[numIds];
        idCode = new String[numIds];
    }

    /**
     * One of the main callbacks from the instrumented StreamIt code.
     * This version is for string values.
     *
     * @param op      The operation being performed.
     * @param id      A static identifier for the line of code calling this.
     * @param val     One of the values taking part in the operation.
     *
     * @return The value of the parameter <pre>val</pre>.
     */
    public static String registerOp(int op, int id, String code, String val) {
        // don't profile string ops, should only be used for debugging
        return val;
    }

    /**
     * Registers a push operation.  Note that this is for a primitive
     * datatype -- will only be called once when pushing arrays.
     */
    public static void registerPop() {
        popTotal++;
    }
    
    /**
     * Registers a push operation.  Note that this is for a primitive
     * datatype -- will only be called once when pushing arrays.
     */
    public static void registerPush() {
        pushTotal++;
    }
    
    /**
     * Called when the program has finished, to print results, etc.
     */
    public static void summarize() {
        try {
            PrintStream out = new PrintStream(new FileOutputStream("countops.java.log"));
            summarize(out);
            out.close();
        } catch (IOException e) {
            e.printStackTrace();
        }

        System.out.println("Operation counts written to countops.java.log.");
    }

    /**
     * Write profile information to <pre>out</pre>.
     */
    private static void summarize(PrintStream out) throws IOException {
        // total ops 
        long opsTotal = floatTotal + intTotal + boolTotal;
        out.println("Total ops:    " + opsTotal);
        out.println("  float ops:  " + floatTotal);
        out.println("  int ops:    " + intTotal);
        out.println("  bool ops:   " + boolTotal);

        // communication
        out.println();
        out.println("Items pushed: " + pushTotal);

        // communication-to-computation ratio
        out.println();
        out.println("Ops / push:   " + (opsTotal / (float)pushTotal));
        out.println("  float+int:  " + ((floatTotal + intTotal) / (float)pushTotal));
        out.println("  bool:       " + (boolTotal / (float)pushTotal));

        // float ops summary
        out.println("\nFloat ops breakdown:");
        for (int i=0; i<NUM_OPS; i++) {
            out.println("  " + OP_TO_NAME[i] + ": " + floatOps[i]);
        }
        // int ops summary
        out.println("\nInt ops breakdown:");
        for (int i=0; i<NUM_OPS; i++) {
            out.println("  " + OP_TO_NAME[i] + ": " + intOps[i]);
        }
        // float ops summary
        out.println("\nBool ops breakdown:");
        for (int i=0; i<NUM_OPS; i++) {
            out.println("  " + OP_TO_NAME[i] + ": " + boolOps[i]);
        }

        // the count by actual statement in the code
        out.println("\nCount for each static operation ID (see .java file for ID's):");
        out.println("  ID        COUNT          % OF TOTAL    CODE (if no arith op, then on RHS of +=, *=, etc) ");
        out.println("------------------------------------------------------------------------------------------");
        for (int i=0; i<idCounts.length; i++) {
            String code = (idCode[i] == null ? "" : idCode[i]);
            out.println("  " + format(i, 10) + format(idCounts[i], 15) + format(100*idCounts[i]/(float)opsTotal, 4) + "          " + code);
        }

        // the sorted count by actual statement in the code
        out.println("\nSorted count for each static operation ID (see .java file for ID's):");
        out.println("  ID        COUNT          % OF TOTAL    CODE (if no arith op, then on RHS of +=, *=, etc) ");
        out.println("------------------------------------------------------------------------------------------");
        // just do selection sort because I had it handy from C++ code
        for (int i=0; i<idCounts.length; i++) {
            int max = 0;
            // find greatest
            for (int j=0; j<idCounts.length; j++) {
                if (idCounts[j] > idCounts[max]) {
                    max = j;
                }
            }
            // print greatest
            String code = (idCode[max] == null ? "" : idCode[max]);
            out.println("  " + format(max, 10) + format(idCounts[max], 15) + format(100*idCounts[max]/(float)opsTotal, 4) + "          " + code);
            // zero-out greatest
            idCounts[max] = -1;
        }
    }

    /**
     * Formats <pre>f</pre> into an <pre>n</pre>-character string.
     */
    private static String format(float f, int n) {
        String str = ""+f;
        int length = str.length();
        for (int j=length; j<n; j++) {
            str += " ";
        }
        return str.substring(0, n);
    }

    /**
     * Formats <pre>i</pre> into an <pre>n</pre>-character string.
     */
    private static String format(int i, int n) {
        String str = ""+i;
        int length = str.length();
        for (int j=length; j<n; j++) {
            str += " ";
        }
        return str;
    }
}
